chainextensions/dapps-staking/impl/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+}
+
+impl TryFrom<u32> for DappsStakingFunc {
+    type Error = DispatchError;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(DappsStakingFunc::CurrentEra),
+            2 => Ok(DappsStakingFunc::UnbondingPeriod),
+            3 => Ok(DappsStakingFunc::EraRewards),
+            4 => Ok(DappsStakingFunc::EraStaked),
+            5 => Ok(DappsStakingFunc::StakedAmount),
+            6 => Ok(DappsStakingFunc::StakedAmountOnContract),
+            7 => Ok(DappsStakingFunc::ReadContractStake),
+            8 => Ok(DappsStakingFunc::Register),
+            9 => Ok(DappsStakingFunc::BondAndStake),
+            10 => Ok(DappsStakingFunc::UnbondAndUnstake),
+            11 => Ok(DappsStakingFunc::WithdrawUnbonded),
+            12 => Ok(DappsStakingFunc::ClaimStaker),
+            13 => Ok(DappsStakingFunc::ClaimDapp),
+            14 => Ok(DappsStakingFunc::SetRewardDestination),
+            15 => Ok(DappsStakingFunc::NominationTransfer),
+            _ => Err(DispatchError::Other(
+                "DappsStakingExtension: Unimplemented func_id",
+            )),
+        }
+    }
+}
+
+// pub struct DappsStakingExtension;
+pub struct DappsStakingExtension<R>(PhantomData<R>);
+
+impl<T: pallet_dapps_staking::Config> ChainExtensionExec<T> for DappsStakingExtension<T> {
+    fn execute_func<E>(func_id: u32, env: Environment<E, InitState>) -> Result<(), DispatchError>
+    where
+        E: Ext<T = T>,
+        <E::T as SysConfig>::AccountId:
+            UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]> + From<[u8; 32]>,
+    {
+        let func_id = DappsStakingFunc::try_from(func_id)?;
+        let mut env = env.buf_in_buf_out();
+
+        match func_id {
+            // DappsStaking - read_current_era()
+            DappsStakingFunc::CurrentEra => {
+                let era_index = pallet_dapps_staking::CurrentEra::<T>::get();
+                env.write(&era_index.encode(), false, None).map_err(|_| {
+                    DispatchError::Other(
+                        "[ChainExtension] DappsStakingExtension failed to write result",
+                    )
+                })?;
+            }
+
+            // DappsStaking - read_unbonding_period()
+            DappsStakingFunc::UnbondingPeriod => {
+                let unbonding_period = T::UnbondingPeriod::get();
+                env.write(&unbonding_period.encode(), false, None)
+                    .map_err(|_| {
+                        DispatchError::Other(
+                            "[ChainExtension] DappsStakingExtension failed to write result",
+                        )
+                    })?;
+            }
+
+            // DappsStaking - read_era_reward()
+            DappsStakingFunc::EraRewards => {
+                let arg: u32 = env.read_as()?;
+                let era_info = pallet_dapps_staking::GeneralEraInfo::<T>::get(arg)
+                    .ok_or(DispatchError::Other("general_era_info call failed"));
```

Comment:
Should `?` be added here?

Reply:
yes it returns a Result
https://github.com/paritytech/substrate/blob/ef890f5598e0d3fd1b6a5fcc8c3322f7570089d1/frame/contracts/src/chain_extension.rs#L369